### DISCUSSION TITLE
CLI console and Script executed

### DISCUSSION THREAD

**Discussion by cybermaus** [sezanzeb/input-remapper#158] [General]
So, after playing with this a little, I find that things do work, especially if I see them described as functionality of this product.

That is, I got it to work when the display manager is started.
When booting into console, I was not able to get it to work, also not when I ensure the service was started and I ran the `key-mapper-control --command start --device "Dell Dell USB Keyboard" --preset Test`

Before finding this, I played a little with evdev driver, and in it;s code that worked as well on CLI as on X. But in the roadmap, I see evdev usuage is for the future, so not yet.

- So is it correct this will not work on non X console? (_Edit: solution see below_)

Also, there is nowhere described (that I can find it) any ability to run shell scripts based on keypresses.

Because that is why I was playing with evdev myself. Ultimately, I want a small numpad keyboard to control processes, mostly by sending CURL commands to other devices.

- So is it correct there is no script calling? (_Edit: solution is custom mod/build, see below_)

**Comment by cybermaus** [comment]
Strike that... Testing again, now it works from the CLI console. 
That means this is useable for me. I can probably either hack in, or properly build in the shell script calls

**Reply by sezanzeb** [reply]
Glad to read that it worked in ttys in the end

> Also, there is nowhere described (that I can find it) any ability to run shell scripts based on keypresses.

> So is it correct there is no script calling?

correct, key-mapper can't do that. It is really just a tool for injecting keyboard events

**Reply by cybermaus** [reply]
well, the jury is still out on the CLI.  It seems the Xorg must be started once, its OK if I fall back into CLI. But if I boot straight into CLI it still does not work, I need to start and stop Xorg for it to work. Or maybe swap from Ctrl-F1 to Ctrl-F7 and back, not sure yet.

As to shell commands: Soon it will be able to do that. At least in a fork of my own. Should I do that in a manner that could be a PR, or does that not fit your plans for this tool? I will program a little more cleanly otherwise.

**Reply by sezanzeb** [reply]
The service runs as root, I'm not sure about the security implications.

If the configured script is supposed to start a gui tool then for it to work in Wayland a daemon has to start that runs as the user that owns the session.

**Reply by sezanzeb** [reply]
> does that not fit your plans for this tool?

There are other tools that do that already, for example one is included in xfce4. I don't think this is in the scope for key-mapper, maybe as a plugin if plugin support ever happens (https://github.com/sezanzeb/key-mapper/issues/80)

**Reply by cybermaus** [reply]
Re: security: Yes, I did check and realise it runs as root. We would need to consider security, but if I am doing it only for myself, on a raspi, I am not too worried. 

Re: GUI: For me, no need for GUI. All it needs to do is turn of/on screen (tvservice -o and -p) and jump to Dahua camera presets using specific curl strings.

Re: other tools: Quite possible that I am re-inventing things that already exists. That is why I stopped messing with evdev myself, and started to look around and found your tool. If you know an even better one, do let me know, it is not impossible I am completely overlooking a famous and obvious one.

**Reply by sezanzeb** [reply]
I unfortunately don't know one that can do it in ttys, yeah.

Maybe this helps: https://dev.to/ecologic/using-keyboard-shortcuts-with-zsh-16b

**Reply by cybermaus** [reply]
Ok, I got it to work with a new macro function exitpgm()

I thought to make it nicely "in-scope" (for as much as I can judge so) with the core-function of your tool: The exit program is "not intended" for system calls, but instead allows to dynamically / externally construct and feed back keystokes and macros. It should return a valid macro string. Of course, I can then also "misuse" the exitpgm for my CURL commands.

However, I am unable to convert the return string into an actual macro. 

I clearly do not understand well how the macro's are loaded, I cannot get the parse() function to work, most likely because I do not provide a proper mapping class. Anyway, it already works for my personal goal, of calling CURL as a result of key presses, so I am giving up on my attempts to do it nicely.

Here is my commit in my own fork, if you are interested and have 5 minutes spare.
commit cf26fff6005b91c804f9a70426077d4db28c0e0d

And how it looks as a json map (so with exitpgm from custom build)
```
{
    "mapping": {
        "1,69,1": "set(dev,cam01).w(10000).ifeq(dev,cam01,set(dev,none))",
        "1,98,1": "set(dev,cam02).w(10000).ifeq(dev,cam02,set(dev,none))",
        "1,55,1": "set(dev,mon01).w(10000).ifeq(dev,mon01,set(dev,none))",
        "1,74,1": "ifeq(dev,mon01,k(KEY_ESC))",
        "1,96,1": "ifeq(dev,mon01,k(KEY_RIGHT))",
        "1,78,1": "ifeq(dev,mon01,k(KEY_ENTER))",
        "1,82,1": "exitpgm(dev,auto)",
        "1,79,1": "exitpgm(dev,preset1)",
        "1,80,1": "exitpgm(dev,preset2)",
        "1,81,1": "exitpgm(dev,preset3)",
        "1,75,1": "exitpgm(dev,preset4)",
        "1,76,1": "exitpgm(dev,preset5)",
        "1,77,1": "exitpgm(dev,preset6)",
        "1,71,1": "exitpgm(dev,preset7)",
        "1,72,1": "exitpgm(dev,preset8)",
        "1,73,1": "exitpgm(dev,preset9)",
        "1,83,1": "exitpgm(dev,toggleblank)"
    }
}
```

**Reply by cybermaus** [reply]
> well, the jury is still out on the CLI. It seems the Xorg must be started once, its OK if I fall back into CLI. But if I boot straight into CLI it still does not work, I need to start and stop Xorg for it to work. Or maybe swap from Ctrl-F1 to Ctrl-F7 and back, not sure yet.

To answer my own question: the service may starts by system init. but in the default build the autoload is driven by X starting. <br/>
To autoload when starting into tty/CLI only, add this to rc.local
```(sleep 2; sudo -u pi key-mapper-control --command autoload) &```
Note the `sudo -u pi` means we are stepping ***down*** from root, needed as the .config directory is located home for user `pi` (on a Raspberry Pi).

**Reply by sezanzeb** [reply]
what is `key-mapper-exit`?

**Reply by cybermaus** [reply]
The fixed name of the exit program, that the user (me, or whoever uses this) needs to write and place in $PATH.
Specifically, for me [this one](https://github.com/cybermaus/key-mapper/blob/my_version/examples/key-mapper-exit.sh)

For me the name seemed to make sense. I realise now that for you it may seem intrusive, claiming a name that clearly belongs in your name space. Sorry.

Edit/Addendum: So the original though was that "key-mapper-exit" would/could return a string in the form of "k(a).k(b)" or something that could be interpreted as a macro. Not that I needed that. But it seemed like something that fitted the key-mapper goals.

**Reply by sezanzeb** [reply]
> I realise now that for you it may seem intrusive, claiming a name that clearly belongs in your name space. Sorry.

No problem. I was just curious what that command is about

> I cannot get the parse() function to work

Lots of examples for the parse function are here: https://github.com/sezanzeb/key-mapper/blob/main/tests/testcases/test_macros.py

**Reply by cybermaus** [reply]
Ok, I got it to work! commit (in my personal fork) e74bebd7adceec8be233d80a1951cfa2529a15f3

And it turns out I have use for it: In most conditions I need to run `curl` commands and even a `systemctl` command. But in one specific condition, I can get away with just the Q key pressed, so I return: `'k(KEY_Q)'`<br/>
And this now works!
[Exit example](https://github.com/cybermaus/key-mapper/blob/my_version/examples/key-mapper-exit.sh) / [key mapping example](https://github.com/cybermaus/key-mapper/blob/my_version/examples/camerakeyboard.json)

Thanks for your help. 
***
BTW: though this is all personal use, I did considered security. I think best way would be to check who is the owner of the exit program, and its [sticky bit](https://en.wikipedia.org/wiki/Sticky_bit). And if sticky bit set, run the subprocess under that user, and if not set, run it as user `nobody` <br/>
This is not implemented though. At this moment, there is the lesser security that any hacker would need to create the exit program in the $PATH of root, which is also already a problem.
***
BTW2: I realise I stubbornly keep calling this an "exit program" where you would call it a (somewhat rudimentary) plugin.
This shows my long experience (or maybe deformation) as IBM midrange programmer, where exit program is the normal terminology.<br/>
Though (in my mind) there is truly a difference. An exit program is hardcoded, the application calls the program if it can find it, thus allowing program to temporary make a detour (an 'exit') from native programming.<br/>
Whereas a plugin system has a external system announce and register itself, in a much more flexible way. Often multiple plugins can register, often they also add themselves to menu classes and screen elements, etc.

**Reply by sezanzeb** [reply]
> And this now works! Thanks for your help.

Great, you're welcome!

I wonder if parsing during runtime is a large bottleneck, but it's certainly an interesting idea I didn't think of yet. To make it faster, you could provide `k(KEY_Q)` as a parameter to the macro function, something like `exitpgm(dev, auto, k(q))`. In that case you could parse the macro already when the injection starts.

> where exit program is the normal terminology

Yeah, I never heard of exit other than in bash and python to close the shell.

> Whereas a plugin system has a external system announce and register itself, in a much more flexible way

Yes, unfortunately I didn't come up with a concept of how that would look like. If you have some ideas you are welcome to throw them into a comment on the issue https://github.com/sezanzeb/key-mapper/issues/80
